### PR TITLE
snv_callers module changes for kfbnl data

### PR DESCRIPTION
**Pull request by ewafula**
### Purpose/implementation Section
* Created modified copies of OpenPBTA snv_callers analysis module scripts, `run_caller_consensus_analysis-kfnbl.sh` and `03-kfnlb-calculate_tmb.R` for KFNBL data.
* Reformatted the **strelka2**, **lancet**, **mutect2**, and **vardict** merged MAF files to work with  OpenPBTA snv_callers analysis module script, `01-setup_db.py` .

#### What scientific question is your analysis addressing?
* SNV consensus analysis using merged multiple SNV callers MAF files.

#### What was your approach?
* Renamed the `variant_id` and `variant_qual` MAF table columns to `vcf_id` and `vcf_qual` for **strelka2**, **lancet**, **mutect2**, and **vardict** merged MAF files to match the required coded sqlite db table schema in the OpenPBTA snv_callers analysis module script, `01-setup_db.py` .
* Performed snv consensus analysis using the **132** common field among all the **1,324** MAF files (**331** files for each caller). There is an additional terminal field (**column 133 == "vcf_pos"**) unique to **22 vardict** MAF files that I excluded in  the consensus analysis to enable python Pandas module  in `01-setup_db.py` script to load the dataframe. 

### Results
* Uploaded snv consensus analysis results to  CAVATICA [**"nbl-x01-wgs-maf-proband/ kfnbl_merged_maf"**](https://cavatica.sbgenomics.com/u/harenzaj/nbl-x01-wgs-maf-proband/files/#q?path=kfnbl_merged_maf) projects.
```
intersect_strelka_mutect_WGS.bed
kfnbl-snv-consensus-mutation.maf.tsv.gz
kfnbl-snv-mutation-tmb-all.tsv
kfnbl-snv-mutation-tmb-coding.tsv
```
 *Replaced reformatted  **strelka2**, **lancet**, **mutect2**, and **vardict** merged MAF files on  CAVATICA [**"nbl-x01-wgs-maf-proband/ kfnbl_merged_maf"**](https://cavatica.sbgenomics.com/u/harenzaj/nbl-x01-wgs-maf-proband/files/#q?path=kfnbl_merged_maf) project.
````
kfnbl-snv-lancet.vep.maf.gz
kfnbl-snv-mutect2.vep.maf.gz
kfnbl-snv-strelka2.vep.maf.gz
kfnbl-snv-vardict.vep.maf.gz
````

### CODE REVIEW

File: analyses/snv-callers/scripts/01-setup_db.py
Comment (jharenza) [inline analyses/snv-callers/scripts/01-setup_db.py:1]: I do not think this change is necessary, as python3 is the default install in docker.
```suggestion
#!/bin/env python
```

**Review by jharenza — COMMENTED** [review]
I think we need to decide how we will add new data starting with this PR :). Adding new scripts with a different prefix is redundant and also doesn't allow us to determine where changes were made. I think I anticipated that we would add additional files to the old scripts or maybe create some parsable options for adding files to the scripts such that we would really change the run script. 

I think adding files per dataset will add up really quickly, so I don't really favor that.

That being said, if we leave this as is for now, then the run script should be added to the Dockerfile [here](https://github.com/PediatricOpenTargets/OpenPBTA-analysis/blob/dev/Dockerfile).

Will also let @kgaonkar6 weigh in as to how best we optimize this repo.

File: analyses/snv-callers/scripts/01-setup_db.py
Comment (ewafula) [reply analyses/snv-callers/scripts/01-setup_db.py]: @jharenza, I have updated the run script,  now  named `run_caller_consensus_analysis.sh` and  the TBM  script `03-calculate-tmb.R` to accept a parsable option for the project acronym prefix that is prepended to all input and output files names hardcoded in the scripts. 
The default python the the docker container is python v2.7.13. The reason why script `01-setup_db.py` is runs successfully is because it is called inside run bash script with python3, not python. I have reverted the shebang back to python.

**Comment by kgaonkar6** [comment]
Thanks for the updated code @ewafula here the input param is a great idea to make it more flexible @jharenza .

I had a few questions and suggestions:
>Renamed the variant_id and variant_qual MAF table columns to vcf_id and vcf_qual for strelka2, lancet, mutect2, and vardict merged MAF files to match the required coded sqlite db table schema in the OpenPBTA snv_callers analysis module script, 01-setup_db.py 

Were the column name changes updated to s3? It seems the download script still has files with `variant_id` and `variant_qual`. 

> Performed snv consensus analysis using the 132 common field among all the 1,324 MAF files (331 files for each caller). There is an additional terminal field (column 133 == "vcf_pos") unique to 22 vardict MAF files that I excluded in the consensus analysis to enable python Pandas module in 01-setup_db.py script to load the dataframe.

I think for now we can remove the vardict files from `analyses/snv-callers/run_caller_consensus_analysis.sh` since there seem to be a discrepancy within the maf files from vardict. We can re-run when we have the updated vardict files ( we don't use vardict in the consensus calling so output files will not be affected)

**Comment by ewafula** [comment]
@kgaonkar6, in response to your question. I already uploaded the merged MAF files with column updates (`variant_id` and `variant_qual`) and other snv consensus analyses results to CAVATICA project [**"nbl-x01-wgs-maf-proband/ kfnbl_merged_maf"**] where @jharenza obtained them last time to upload to S3 bucket. I don't have access to upload data to the OpenPBTA S3 bucket where OpenTargets data is hosted.

**Comment by kgaonkar6** [comment]
oh ok I'll update the files on s3.

**Comment by kgaonkar6** [comment]
The updated files are now in s3, and the run script successfully created consensus maf and tmb files using the updated strelka2,mutect2 and lancet and subset of vardict files.

I believe we will also have to re-create plots with compare_snv_callers_plots.Rmd for each new project ( this is not run by default since it takes a long time).  Currently the output files are coded as "pbta"* maybe it is also a good idea to add a param to  the notebook within the YAML header of the document which can then read in the project_acronym.
```
params:
  project_acronym:
    label: "project acronymn to use as prefix for output files"
    value: kfnbl
    input: string
```

**Comment by ewafula** [comment]
@kgaonkar6, I have made the changes you recommended:
* amended the `compare_snv_callers_plots.Rmd` script to include `project_acronym` parameter to prefix project input and output files
* changed the calling of `compare_snv_callers_plots.Rmd` script in the bash `run_caller_consensus_analysis.sh` script to include the project acronym prefix parameter and tag the rendered notebook html output file name with the project acronym
* edited the `README.md` file to show how to run the bash `run_caller_consensus_analysis.sh` script with the parameter option for the project acronym

**New files in the updated PR:**
```
analyses/snv-callers/compare_snv_callers_plots-kfnbl.nb.html
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-frac_base_change_plot.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-upset-plot-not-transcipts-only.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-upset-plot-transcipts-only.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-upset-plot.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-vaf_cor_matrix.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-vaf_distribution_plot.png
analyses/snv-callers/plots/kfnbl-caller-comparison/kfnbl-variant_classification_plot.png
```

**Review by kgaonkar6 — APPROVED** [review]
Thanks for the update! LGTM!

**Comment by ewafula** [comment]
> Just wanted to mention, I downloaded the consensus file from "nbl-x01-wgs-maf-proband/ kfnbl_merged_maf" and uploaded to the s3 bucket you can re-run the download-data.sh to get the most updated files for any downstream analysis.

@kgaonkar6, I already downloaded updated s3 bucket files and used the consensus file for compare snv plotting. Should I merge this PR or wait for everyone to approve first? Is there any other change required for the consensus snv module?

**Comment by kgaonkar6** [comment]
The changes look good to me, I think we should wait for @jharenza approval before we can merge this PR.

**Review by migbro — APPROVED** [review]
Seems good. The repo only seems to need one approval instead of teo by default

**Comment by jharenza** [comment]
Hi @ewafula - I wanted to wait on merging this until we have v2 data in here. Once @zhangb1 is finished with the INDEL renorm/merging, we can then update this code and get the final consensus maf.

**Comment by jharenza** [comment]
@ewafula now that #5 is in, if you can download that data and rerun this, i think we can get this merged.

**Comment by ewafula** [comment]
@jharenza, have to spend a bit more time figuring out why the snv_callers module script `0-set-up_db.py` is not running okay. I see the new merged MAFs for all callers have the`MQ` column but somehow the script doesn't recognize it. 
```
Reading file data/kfnbl-snv-strelka2.vep.maf.gz to table 'strelka'.
Traceback (most recent call last):
  File "analyses/snv-callers/scripts/01-setup_db.py", line 309, in <module>
    chunk.to_sql(table_name, con, if_exists='append')
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/generic.py", line 2712, in to_sql
    method=method,
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/sql.py", line 518, in to_sql
    method=method,
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/sql.py", line 1750, in to_sql
    table.insert(chunksize, method)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/sql.py", line 756, in insert
    exec_insert(conn, keys, chunk_iter)
  File "/usr/local/lib/python3.5/dist-packages/pandas/io/sql.py", line 1468, in _execute_insert
    conn.executemany(self.insert_statement(), data_list)
sqlite3.OperationalError: table strelka has no column named MQ
```

**Comment by ewafula** [comment]
@jharenza, there are four new columns `MQ`, `MQ0`, `QSI`, and `HotSpotAllele`. They were not coded in the data schema in the `01-setup_db.py` script for creating the Sqlite3 database. But also the data values in these columns are not uniform within/among all the four callers. I made them all "TEXT" datatype. Please look at the first few lines of each caller MAF.
`zless kfnbl-snv-vardict.vep.maf.gz  | cut -f 113-116 | head -n 50`

Still trying to solve the error when loading vardict MAF to the database. Others callers loaded successfully.
Vardict might needs a bit of careful cleaning. I'll come back to it after completing the TARGET histologies. Unless someone else has previously encountered such an error before while working the snv-callers module and fixed it.
```
Reading file data/kfnbl-snv-vardict.vep.maf.gz to table 'vardict'.
Traceback (most recent call last):
  File "pandas/_libs/lib.pyx", line 1897, in pandas._libs.lib.maybe_convert_numeric
ValueError: Unable to parse string "+"
...
File "pandas/_libs/lib.pyx", line 1934, in pandas._libs.lib.maybe_convert_numeric
ValueError: Unable to parse string "+" at position 0
```

**Comment by migbro** [comment]
> @jharenza, there are four new columns `MQ`, `MQ0`, `QSI`, and `HotSpotAllele`. They were not coded in the data schema in the `01-setup_db.py` script for creating the Sqlite3 database. But also the data values in these columns are not uniform within/among all the four callers. I made them all "TEXT" datatype. Please look at the first few lines of each caller MAF.
> `zless kfnbl-snv-vardict.vep.maf.gz | cut -f 113-116 | head -n 50`
> 
> Still trying to solve the error when loading vardict MAF to the database. Others callers loaded successfully.
> Vardict might needs a bit of careful cleaning. I'll come back to it after completing the TARGET histologies. Unless someone else has previously encountered such an error before while working the snv-callers module and fixed it.
> 
> ```
> Reading file data/kfnbl-snv-vardict.vep.maf.gz to table 'vardict'.
> Traceback (most recent call last):
>   File "pandas/_libs/lib.pyx", line 1897, in pandas._libs.lib.maybe_convert_numeric
> ValueError: Unable to parse string "+"
> ...
> File "pandas/_libs/lib.pyx", line 1934, in pandas._libs.lib.maybe_convert_numeric
> ValueError: Unable to parse string "+" at position 0
> ```

Hi Eric. I think I can add some context as to why that is. Each caller has a handful `INFO` fields in the vcf that are native to the caller itself, some of which I felt could be useful to downstream users. These were included in their respective caller MAF files. If this is causing issues and/or you'd rather not code for each situation, they are tacked on as the last columns. So, you can safely ignore columns between `vcfpos` (I think that's the last common column) and `HotSpotAllele`. Hope this helps and please let me know if there are more questions/concerns!

**Comment by jharenza** [comment]
@ewafula have you merged the latest changes from dev into your branch? I think that should take care of it since I pulled these changes from the ALSF repo.

**Comment by kgaonkar6** [comment]
@ewafula can you please update your forked repo to `PediatricOpenTargets/OpenPBTA-analysis`? We did update the `0-set-up_db.py` to ignore those extra columns in the new MAF formats.


Have you set up `PediatricOpenTargets/OpenPBTA-analysis` as a remote upstream repository? You can follow these instructions for configuring a remote that points to an upstream repository here if you have not done this already:
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/configuring-a-remote-for-a-fork

Then we would basically have to get latest changes from upstream and merge to your forked repo. This is psuedo git commands which I think will help:

```
# fetch the changes from the PediatricOpenTargets/OpenPBTA-analysis repository
git fetch upstream

# checkout your own master branch
git checkout master

# merge the upstream/master branch into your own master branch
git merge upstream/master

# push the changes to your master branch to the remote repository
git push

# Since we want to use the `dev` branch as default in your forked repo as well
git checkout dev 
git merge master
```

**Comment by ewafula** [comment]
@migbro, @kgaonkar6, @jharenza, my local dev branch is now update. But problem is still there when Pandas is reading the the vardict MAF file. In the `0-set-up_db.py` script, the `HotSpotAllele` column is coded as of type "INTEGER"  but the values are  of type "FLOAT".  Pandas is having a problem that. There is also an additional last headerless column is in the vardict MAF not present in the MAFs of the other three callers.  When I remove the `HotSpotAllele` column with float values in the vardict MAF and rename the headerless column as`HotSpotAllele`, the `0-set-up_db.py` script runs okay. Please take a look and at the vardict MAF and advise.

**Comment by jharenza** [comment]
> @migbro, @kgaonkar6, @jharenza, my local dev branch is now update. But problem is still there when Pandas is reading the the vardict MAF file. In the `0-set-up_db.py` script, the `HotSpotAllele` column is coded as of type "INTEGER"  but the values are  of type "FLOAT".  Pandas is having a problem that. There is also an additional last headerless column is in the vardict MAF not present in the MAFs of the other three callers.  When I remove the `HotSpotAllele` column with float values in the vardict MAF and rename the headerless column as`HotSpotAllele`, the `0-set-up_db.py` script runs okay. Please take a look and at the vardict MAF and advise. 

Also adding @zhangb1 here in case we need to tweak the workflow upstream.

**Comment by migbro** [comment]
I took a look and it's just a simple mistake in the merge file. It is missing the column header `SSF` before `HotSpotAllele`.
If you add that @zhangb1 and re-upload, should fix the problem 😀

**Comment by jharenza** [comment]
thanks @migbro - @zhangb1 please put this in a v3 folder and we will do a new release.

**Comment by kgaonkar6** [comment]
Hi @ewafula, I think the only task left before this PR can be merged is to up-date this remote branch. Will you push your updated local `snv_callers_consensus_analysis` branch after you have merged with the up-to-date `dev` branch code?

**Comment by ewafula** [comment]
@kgaonkar6, pushed updated local `snv_callers_consensus_analysis` branch to this remote branch.

**Pull request merged**
